content/blog/2020-10-tokio-0-3.md

```diff
@@ -0,0 +1,129 @@
+
+We have reduced the amount of feature flags needed to use the core features of Tokio. The `dns`, `tcp`, `udp`,  and `uds` feature flags are collapsed into a single `net` feature flag. We have also combined `rt-core` and `rt-util` into a single `rt` feature flag. This `rt` feature flag now encompasses everything to execute futures except for the multi-threaded runtime which now lives under a `rt-multi-thread` feature flag (previously was called `rt-threaded`).
+
+- `dns`, tcp`, `udp` and `uds` -> `net`
+- `rt-util` and `rt-core` -> `rt`
+- `rt-threaded` -> `rt-multi-threaded`
+
+## Runtime and Builder refactor
+
+The Tokio team has improved the runtime `Builder` to provide better consistency accross feature flag permutations and to be more mis-use resistant. To achieve this, we have added two new constructors to the `Builder` type to support building both variations of the Tokio runtime. Additionally, Tokio will no longer choose the runtime variant based on feature flags when using `Runtime::new` but instead will default to the multi-threaded runtime and is only available under the `rt` feature flag. In addition, we’ve renamed the `core_threads` builder method to the more accurate `worker_threads`. 
+
+We have also refactored the runtime module. Tokio 0.2 exposed two types to interact with the Tokio runtime: `Runtime` and `Handle`. `Runtime` has a `block_on` method that required `&mut self`, whereas `Handle` has a `block_on` method that required `&self`. With Tokio 0.3, we collapsed `Runtime` and `Handle` into a single `Runtime`. Tokio 0.3’s `Runtime`‘s `block_on` now accepts `&self`, which supports concurrent calls to `block_on`. This also means `Runtime` implements `Send` and `Sync` and can therefore be stored in an `Arc`.
+
+An example configuring a multi-threaded runtime with 6 worker threads:
+
+```rust
+use tokio::runtime::Builder;
+
+let rt = Builder::new_multi_thread()
+    .worker_threads(6)
+    .enable_all()
+    .build()
+    .unwrap();
+
+rt.block_on(async {
+    println!("Hello world!");
+});
+```
+
+## Methods changed to `&self`
+
+We have overhauled how we handle wakers within our socket types to allow concurrent calls via `&self` rather than `&mut self`. To achieve this we had to rewrite how we handle wakers internally. The problem comes with how `poll_*` based methods work. Under the futures contract it states that a `poll_*` fn should only store the _last_ waker that it was called with. 
+
+> Note that on multiple calls to poll, only the Waker from the Context passed to the most recent call should be scheduled to receive a wakeup.
+
+From [`Future`](doc.rust-lang.org/std/future/trait.Future.html)
+
+This means that these `poll_*` can not be called from or woken concurrently. To make `&self` work for our `async fn` on our sockets types we refactored our io driver to use an internal instrusive linked-list to store the wakers. By doing this, it allows our socket types to accept `&self` rather than` &mut self` calls. 
```

Comment:
```suggestion
This means that these `poll_*` methods cannot be called or woken concurrently. To make `&self` work for our `async fn` on our sockets types, we refactored our io driver to use an internal instrusive linked-list to store the wakers. By doing this, our socket types can accept `&self` rather than` &mut self` when called. 
```